torchvision/datasets/inaturalist.py

```diff
@@ -0,0 +1,136 @@
+            transform: Optional[Callable] = None,
+            target_transform: Optional[Callable] = None,
+    ) -> None:
+        super(INaturalist, self).__init__(root,
+                                          transform=transform,
+                                          target_transform=target_transform)
+        if not isinstance(target_type, list):
+            target_type = [target_type]
+        self.target_type = [verify_str_arg(t, "target_type", ("full", *CATEGORIES))
+                            for t in target_type]
+
+        if not os.path.exists(self.root):
+            raise RuntimeError(f'Dataset not found in {self.root}. You need to download it.')
+        self.all_categories = sorted(os.listdir(self.root))
+        if (not self.all_categories):
+            raise RuntimeError(f'Dataset not found in {self.root}. You need to download it.')
+
+        # map: category type -> name of category -> index
+        self.categories_index: Dict[str, Dict[str, int]] = {
+            k: {} for k in CATEGORIES
+        }
+        # list indexed by category id, containing mapping from category type -> index
+        self.categories_map: List[Dict[str, int]] = []
+
+        # index of all files: (full category id, filename)
+        self.index: List[Tuple[int, str]] = []
+
+        for dir_index, dir_name in enumerate(self.all_categories):
+            pieces = dir_name.split('_')
+            if len(pieces) != 8:
+                raise RuntimeError(f'Unexpected category name {dir_name}, wrong number of pieces')
+            if pieces[0] != f'{dir_index:05d}':
+                raise RuntimeError(f'Unexpected category id {pieces[0]}, expecting {dir_index:05d}')
+            cat_map = {}
+            for cat, name in zip(CATEGORIES, pieces[1:7]):
+                if name in self.categories_index[cat]:
+                    cat_id = self.categories_index[cat][name]
+                else:
+                    cat_id = len(self.categories_index[cat])
+                    self.categories_index[cat][name] = cat_id
+                cat_map[cat] = cat_id
+            self.categories_map.append(cat_map)
+
+            files = os.listdir(os.path.join(self.root, dir_name))
+            for fname in files:
+                self.index.append((dir_index, fname))
+
+    def __getitem__(self, index: int) -> Tuple[Any, Any]:
+        """
+        Args:
+            index (int): Index
+
+        Returns:
+            tuple: (image, target) where the type of target specified by target_type.
+        """
+
+        cat_id, fname = self.index[index]
+        img = Image.open(os.path.join(self.root,
+                                      self.all_categories[cat_id],
+                                      fname))
+
+        target: Any = []
+        for t in self.target_type:
+            if t == "full":
+                target.append(cat_id)
+            else:
+                target.append(self.categories_map[cat_id][t])
```

Comment:
>  By returning the different types as different keys in the return dictionary

That is what I had in mind. Since we will be longer restricted by the output "size", we can simply always return everything. Only exception I would make here is if the additional information increases the load time, e.g. loading extra segmentation masks if the dataset normally only supplies bounding boxes.